src/app/server/JointFabricDatastore.cpp

```diff
@@ -0,0 +1,284 @@
+
+    for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->MarkNodeListChanged();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR JointFabricDatastore::UpdateNode(NodeId nodeId, const CharSpan & friendlyName)
+{
+    for (auto & entry : mNodeInformationEntries)
+    {
+        if (entry.nodeID == nodeId)
+        {
+            entry.Set(MakeOptional(friendlyName));
+
+            for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+            {
+                listener->MarkNodeListChanged();
+            }
+
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR JointFabricDatastore::RemoveNode(NodeId nodeId)
+{
+    for (auto it = mNodeInformationEntries.begin(); it != mNodeInformationEntries.end(); ++it)
+    {
+        if (it->nodeID == nodeId)
+        {
+            mNodeInformationEntries.erase(it);
+
+            for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+            {
+                listener->MarkNodeListChanged();
+            }
+
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR JointFabricDatastore::RefreshNode(NodeId nodeId)
+{
+    // 1. && 2.
+    ReturnErrorOnFailure(SetNode(nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum::kPending));
+
+    // 3. TODO: Read the PartsList of the Descriptor cluster from the Node.
+
+    // 4. TODO
+
+    // 5. TODO
+
+    // 6.
+    ReturnErrorOnFailure(SetNode(nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum::kCommitted));
+
+    for (Listener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+    {
+        listener->MarkNodeListChanged();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR JointFabricDatastore::SetNode(NodeId nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum state)
+{
+    size_t index = 0;
+    ReturnErrorOnFailure(IsNodeIDInDatastore(nodeId, index));
+    mNodeInformationEntries[index].commissioningStatusEntry.state = state;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR JointFabricDatastore::IsNodeIDInDatastore(NodeId nodeId, size_t & index)
+{
+    for (auto & entry : mNodeInformationEntries)
+    {
+        if (entry.nodeID == nodeId)
+        {
+            index = static_cast<size_t>(&entry - &mNodeInformationEntries[0]);
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_KEY_NOT_FOUND;
+}
+
+CHIP_ERROR
+JointFabricDatastore::AddGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet)
+{
+    VerifyOrReturnError(IsGroupKeySetEntryPresent(groupKeySet.groupKeySetID) == false, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mGroupKeySetList.size() < kMaxNodes, CHIP_ERROR_NO_MEMORY);
```

Comment:
This check should be done against `kMaxGroupKeySet`

Reply:
Addressed in the latest commit.